{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "Complement =  {\"A\":\"T\", \"T\":\"A\",\"G\":\"C\",\"C\":\"G\"}\n",
    "def reverse_complement(seq):\n",
    "    \"\"\" Swaps Guanine with cytosine and Adenine with thymine and vice versa. Then reverses the string\"\"\"\n",
    "    string = ''\n",
    "    for nuc in seq:\n",
    "        string += Complement[nuc]\n",
    "    return string[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open(r\"D:\\Summers\\BSE322\\Algorithms-DNA-Sequencing-master\\1_week\\lambda_virus.fa\",'r')\n",
    "file.readline().replace(\"\\n\",\"\")\n",
    "genome = file.read().replace(\"\\n\",\"\")\n",
    "#print(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'G': 12820, 'A': 12334, 'T': 11986, 'C': 11362})\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "import collections\n",
    "count = Counter(genome)\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive(seq, sub_seq):\n",
    "    \"\"\" searches for all occurences of subsequence(text) in a given sequence\"\"\"\n",
    "    occurences = []\n",
    "    for i in range(0,len(seq) - len(sub_seq) + 1):\n",
    "        match = True\n",
    "        for j in range(len(sub_seq)):\n",
    "            if not seq[i+j] == sub_seq[j]:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurences.append(i)\n",
    "    return occurences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "question1 = '''1. How many times does AGGT or its reverse complement (ACCT) occur in the lambda virus genome?\n",
    "E.g. if AGGT occurs 10 times and ACCT occurs 12 times, you should report 22.'''\n",
    "#print(question1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "306\n"
     ]
    }
   ],
   "source": [
    "n = len(naive(genome, 'AGGT')) + len(naive(genome, \"ACCT\"))\n",
    "print(n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "question2 = '''2. How many times does TTAA or its reverse complement occur in the labda virus genome?\n",
    "Hint: TTAA and its reverse complement are equal, so remember to not double count'''\n",
    "#print(question2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "195\n"
     ]
    }
   ],
   "source": [
    "n = len(naive(genome, 'TTAA'))\n",
    "print(n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "question3 = '''What is the offset of the leftmost occurrence of ACTAAGT or its reverse complement\n",
    "in the Lambda virus genome? E.g. if the leftmost occurrence of ACTAAGT is at offset 40 (0-based)\n",
    "and the leftmost occurrence of its reverse complement ACTTAGT is at offset 29, then report 29.'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26028\n"
     ]
    }
   ],
   "source": [
    "occurences1 = naive(genome, 'ACTAAGT')\n",
    "occurences2 = naive(genome, reverse_complement('ACTAAGT'))\n",
    "ans = min(occurences1[0],occurences2[0])\n",
    "print(ans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "question4 = '''What is the offset of the leftmost occurrence of AGTCGA or its reverse complement\n",
    "in the Lambda virus genome?'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "occurences1 : 18005 occurences2 : 450\n",
      "450\n"
     ]
    }
   ],
   "source": [
    "occurences1 = naive(genome, 'AGTCGA')\n",
    "occurences2 = naive(genome, reverse_complement('AGTCGA'))\n",
    "print(\"occurences1 : %d occurences2 : %d\" % (occurences1[0],occurences2[0]))\n",
    "ans = min(occurences1[0],occurences2[0])\n",
    "print(ans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "question5 = '''As we will discuss, sometimes we would like to find approximate matches for P in T.\n",
    "That is, we want to find occurrences with one or more differences.\n",
    "\n",
    "For Questions 5 and 6, make a new version of the naive function called naive_2mm\n",
    "that allows up to 2 mismatches per occurrence. Unlike for the previous questions,\n",
    "do not consider the reverse complement here. We're looking for approximate matches for P itself,\n",
    "not its reverse complement.\n",
    "\n",
    "For example, ACTTTA occurs twice in ACTTACTTGATAAAGT, once at offset 0 with 2 mismatches,\n",
    "and once at offset 4 with 1 mismatch. So naive_2mm(’ACTTTA’,’ACTTACTTGATAAAGT’)\n",
    "should return the list [0,4].\n",
    "\n",
    "How many times does TTCAAGCC occur in the Lambda virus genome when allowing up to 2 mismatches?'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 4]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def naive_modified(seq,sub_seq,n):\n",
    "    \"\"\"allows n mismatches while matching\"\"\"\n",
    "    occurences = []\n",
    "    for i in range(0,len(seq) - len(sub_seq) + 1):\n",
    "        match = True\n",
    "        temp_mis = 0\n",
    "        for j in range(len(sub_seq)):\n",
    "            if not seq[i+j] == sub_seq[j]:\n",
    "                temp_mis += 1\n",
    "            if temp_mis > n:\n",
    "                match = False\n",
    "                break                \n",
    "        if match:\n",
    "            occurences.append(i)\n",
    "    return occurences\n",
    "#naive_modified( 'ACTTACTTGATAAAGT','ACTTTA', 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "191\n"
     ]
    }
   ],
   "source": [
    "ans = len(naive_modified(genome, \"TTCAAGCC\",2))\n",
    "print(ans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "question6 = '''What is the offset of the leftmost occurrence of AGGAGGTT\n",
    "in the Lambda virus genome when allowing up to 2 mismatches?'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49\n"
     ]
    }
   ],
   "source": [
    "print(naive_modified(genome,\"AGGAGGTT\",2)[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "question7 = '''Finally, download and parse the provided FASTQ file containing real DNA sequencing reads\n",
    "derived from a human:\n",
    "\n",
    "https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq\n",
    "\n",
    "Note that the file has many reads in it and you should examine all of them together when answering this question.\n",
    "The reads are taken from this study:\n",
    "\n",
    "Ajay, S. S., Parker, S. C., Abaan, H. O., Fajardo, K. V. F., & Margulies, E. H. (2011). Accurate\n",
    "\n",
    "and comprehensive sequencing of personal genomes. Genome research, 21(9), 1498-1505.\n",
    "\n",
    "This dataset has something wrong with it; one of the sequencing cycles is poor quality.\n",
    "\n",
    "Report which sequencing cycle has the problem. Remember that a sequencing cycle corresponds\n",
    "to a particular offset in all the reads. For example, if the leftmost read position seems\n",
    "to have a problem consistently across reads, report 0. If the fourth position from the left has the problem,\n",
    "report 3. Do whatever analysis you think is needed to identify the bad cycle.\n",
    "It might help to review the \"Analyzing reads by position\" video.'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "file2 = file = open(r\"D:\\Summers\\BSE322\\Algorithms-DNA-Sequencing-master\\1_week\\ERR037900_1.first1000.fastq\",'r')\n",
    "#print(file2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fastq(file_name):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    while True:\n",
    "        file_name.readline()  #skip name line\n",
    "        seq = file_name.readline().rstrip() #read base sequence\n",
    "        file_name.readline()  #skip placeholder line\n",
    "        qual = file_name.readline().rstrip() #base quality line\n",
    "        if len(seq) == 0:\n",
    "            break\n",
    "        sequences.append(seq)\n",
    "        qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "sequences,qualities = read_fastq(file2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 17723, 0, 2, 11, 11, 28, 23, 55, 100, 111, 86, 174, 185, 272, 317, 259, 390, 1523, 2782, 762, 286, 413, 403, 538, 351, 694, 971, 777, 1024, 1449, 1341, 1312, 1916, 2233, 3025, 4043, 6640, 45696, 2074, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "def Q_to_phred33(Q):\n",
    "    '''Turn Q into Phred+33 ASCII-­‐encoded quality'''\n",
    "    return chr(Q + 33)          # converts character to integer according to ASCII table\n",
    "def phred33_to_Q(qual):\n",
    "    '''Turn Phred+33 ASCII-encoded quality into Q'''\n",
    "    return ord(qual) - 33       # converts integer to character according to \n",
    "#print(phred33_to_Q('#'))\n",
    "def create_hist(qualities):     #create histogram to count the frequency of each quality score\n",
    "    hist = [0]*50\n",
    "    for qual in qualities:\n",
    "        for i in qual:\n",
    "            hist[phred33_to_Q(i)] += 1\n",
    "    return hist\n",
    "h = create_hist(qualities)\n",
    "print(h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAD4CAYAAADsKpHdAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAPmElEQVR4nO3dX4xc5X3G8e8TmxLUFMKfBSGv26XCqgKoIcKiluhFitPilijmAqSNlOALV5YQkYiUKjK5iVLJkrkJKVJBQiHCkDRgkaRYiVBrmaC0EjJZElJjHMSqULBsYScQQi6gMvn1Yt5Vx8t6d3aZ9axnvh9pdM75zXnPvi+I88x5z8whVYUkSR8adAckSSuDgSBJAgwESVJjIEiSAANBktSsHnQHluqiiy6qiYmJQXdDks4ozz777K+qamyu987YQJiYmGBqamrQ3ZCkM0qS/znVe04ZSZIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkoAz+JfKkkbPxPYfva/2ys4bB9CT4eQVgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSgEUEQpJVSX6e5Idt+4Ike5O81Jbnd+17Z5LpJC8muaGrfk2SA+29e5Kk1c9O8mir708y0ccxSpJ6sJgrhDuAQ13b24F9VbUO2Ne2SXIFMAlcCWwC7k2yqrW5D9gGrGuvTa2+FXizqi4H7gbuWtJoJElL1lMgJBkHbgS+2VXeDOxq67uAm7rqj1TVu1X1MjANXJvkUuDcqnq6qgp4aFabmWM9BmycuXqQJJ0evV4hfAP4MvD7rtolVXUUoC0vbvU1wGtd+x1utTVtfXb9pDZVdQJ4C7iw10FIkj64BQMhyaeBY1X1bI/HnOuTfc1Tn6/N7L5sSzKVZOr48eM9dkeS1IterhCuAz6T5BXgEeD6JN8GXm/TQLTlsbb/YWBtV/tx4Eirj89RP6lNktXAecAbsztSVfdX1fqqWj82NtbTACVJvVkwEKrqzqoar6oJOjeLn6yqzwF7gC1tty3A4219DzDZvjl0GZ2bx8+0aaW3k2xo9wdundVm5lg3t7/xvisESdLyWf0B2u4EdifZCrwK3AJQVQeT7AZeAE4At1fVe63NbcCDwDnAE+0F8ADwcJJpOlcGkx+gX5KkJVhUIFTVU8BTbf3XwMZT7LcD2DFHfQq4ao76O7RAkSQNhr9UliQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEtBDICT5cJJnkvwiycEkX2v1C5LsTfJSW57f1ebOJNNJXkxyQ1f9miQH2nv3JEmrn53k0Vbfn2RiGcYqSZpHL1cI7wLXV9XHgauBTUk2ANuBfVW1DtjXtklyBTAJXAlsAu5Nsqod6z5gG7CuvTa1+lbgzaq6HLgbuOuDD02StBgLBkJ1/K5tntVeBWwGdrX6LuCmtr4ZeKSq3q2ql4Fp4NoklwLnVtXTVVXAQ7PazBzrMWDjzNWDJOn06OkeQpJVSZ4DjgF7q2o/cElVHQVoy4vb7muA17qaH261NW19dv2kNlV1AngLuHCOfmxLMpVk6vjx4z0NUJLUm54Coareq6qrgXE6n/avmmf3uT7Z1zz1+drM7sf9VbW+qtaPjY0t0GtJ0mIs6ltGVfUb4Ck6c/+vt2kg2vJY2+0wsLar2ThwpNXH56if1CbJauA84I3F9E2S9MH08i2jsSQfbevnAJ8CfgnsAba03bYAj7f1PcBk++bQZXRuHj/TppXeTrKh3R+4dVabmWPdDDzZ7jNIkk6T1T3scymwq31T6EPA7qr6YZKngd1JtgKvArcAVNXBJLuBF4ATwO1V9V471m3Ag8A5wBPtBfAA8HCSaTpXBpP9GJwkqXcLBkJV/RfwiTnqvwY2nqLNDmDHHPUp4H33H6rqHVqgSJIGw18qS5IAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1CwZCkrVJfpzkUJKDSe5o9QuS7E3yUlue39XmziTTSV5MckNX/ZokB9p79yRJq5+d5NFW359kYhnGKkmaRy9XCCeAL1XVx4ANwO1JrgC2A/uqah2wr23T3psErgQ2AfcmWdWOdR+wDVjXXptafSvwZlVdDtwN3NWHsUmSFmHBQKiqo1X1s7b+NnAIWANsBna13XYBN7X1zcAjVfVuVb0MTAPXJrkUOLeqnq6qAh6a1WbmWI8BG2euHiRJp8ei7iG0qZxPAPuBS6rqKHRCA7i47bYGeK2r2eFWW9PWZ9dPalNVJ4C3gAvn+PvbkkwlmTp+/Phiui5JWkDPgZDkI8D3gC9W1W/n23WOWs1Tn6/NyYWq+6tqfVWtHxsbW6jLkqRF6CkQkpxFJwy+U1Xfb+XX2zQQbXms1Q8Da7uajwNHWn18jvpJbZKsBs4D3ljsYCRJS9fLt4wCPAAcqqqvd721B9jS1rcAj3fVJ9s3hy6jc/P4mTat9HaSDe2Yt85qM3Osm4En230GSdJpsrqHfa4DPg8cSPJcq30F2AnsTrIVeBW4BaCqDibZDbxA5xtKt1fVe63dbcCDwDnAE+0FncB5OMk0nSuDyQ82LEnSYi0YCFX1n8w9xw+w8RRtdgA75qhPAVfNUX+HFiiSpMHwl8qSJMBAkCQ1BoIkCTAQJElNL98yEjCx/Ucnbb+y88YB9USSlodXCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJKAHgIhybeSHEvyfFftgiR7k7zUlud3vXdnkukkLya5oat+TZID7b17kqTVz07yaKvvTzLR5zFKknrQyxXCg8CmWbXtwL6qWgfsa9skuQKYBK5sbe5Nsqq1uQ/YBqxrr5ljbgXerKrLgbuBu5Y6GEnS0i0YCFX1E+CNWeXNwK62vgu4qav+SFW9W1UvA9PAtUkuBc6tqqerqoCHZrWZOdZjwMaZqwdJ0umz1HsIl1TVUYC2vLjV1wCvde13uNXWtPXZ9ZPaVNUJ4C3gwrn+aJJtSaaSTB0/fnyJXZckzaXfN5Xn+mRf89Tna/P+YtX9VbW+qtaPjY0tsYuSpLksNRBeb9NAtOWxVj8MrO3abxw40urjc9RPapNkNXAe75+ikiQts6UGwh5gS1vfAjzeVZ9s3xy6jM7N42fatNLbSTa0+wO3zmozc6ybgSfbfQZJ0mm0eqEdknwX+CRwUZLDwFeBncDuJFuBV4FbAKrqYJLdwAvACeD2qnqvHeo2Ot9YOgd4or0AHgAeTjJN58pgsi8jkyQtyoKBUFWfPcVbG0+x/w5gxxz1KeCqOerv0AJFkjQ4/lJZkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBBgIkqRmwUdXSNLpNrH9Rydtv7LzxgH1ZLR4hSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUuOjKyQNxOzHU4CPqBg0rxAkSYCBIElqDARJEmAgSJIabypLWlbePD5zGAgaav6PVqTeOWUkSQIMBElS45SRpL4Y5L0C71P0h4EgaVE8+Q4vp4wkSYBXCBoCfmLtjf+ctBADQRpxpwoKA2T0GAjSCraUk7Inci2VgaCRNAwnTX90p35bMYGQZBPwT8Aq4JtVtXPAXZKWbLGBMwwBpTPfigiEJKuAfwb+GjgM/DTJnqp6YbA903I6k06a/Zpn98SvlWxFBAJwLTBdVf8NkOQRYDNgIPTJfCeiU009LPYkOAwnTU/YGmWpqkH3gSQ3A5uq6u/b9ueBv6iqL8zabxuwrW3+GfBiH/78RcCv+nCcM4ljHg2OeTQsdsx/UlVjc72xUq4QMkftfUlVVfcD9/f1DydTVbW+n8dc6RzzaHDMo6GfY14pv1Q+DKzt2h4HjgyoL5I0klZKIPwUWJfksiR/AEwCewbcJ0kaKStiyqiqTiT5AvBvdL52+q2qOnia/nxfp6DOEI55NDjm0dC3Ma+Im8qSpMFbKVNGkqQBMxAkScAIB0KSTUleTDKdZPug+7NcknwrybEkz3fVLkiyN8lLbXn+IPvYT0nWJvlxkkNJDia5o9WHecwfTvJMkl+0MX+t1Yd2zDOSrEry8yQ/bNujMOZXkhxI8lySqVbry7hHMhC6HpXxt8AVwGeTXDHYXi2bB4FNs2rbgX1VtQ7Y17aHxQngS1X1MWADcHv7dzvMY34XuL6qPg5cDWxKsoHhHvOMO4BDXdujMGaAv6qqq7t+f9CXcY9kIND1qIyq+l9g5lEZQ6eqfgK8Mau8GdjV1ncBN53OPi2nqjpaVT9r62/TOVmsYbjHXFX1u7Z5VnsVQzxmgCTjwI3AN7vKQz3mefRl3KMaCGuA17q2D7faqLikqo5C5wQKXDzg/iyLJBPAJ4D9DPmY29TJc8AxYG9VDf2YgW8AXwZ+31Ub9jFDJ+z/Pcmz7XE+0Kdxr4jfIQxAT4/K0JkryUeA7wFfrKrfJnP9Kx8eVfUecHWSjwI/SHLVgLu0rJJ8GjhWVc8m+eSAu3O6XVdVR5JcDOxN8st+HXhUrxBG/VEZrye5FKAtjw24P32V5Cw6YfCdqvp+Kw/1mGdU1W+Ap+jcNxrmMV8HfCbJK3SmfK9P8m2Ge8wAVNWRtjwG/IDOFHhfxj2qgTDqj8rYA2xp61uAxwfYl75K51LgAeBQVX29661hHvNYuzIgyTnAp4BfMsRjrqo7q2q8qibo/Pf7ZFV9jiEeM0CSP0zyRzPrwN8Az9OncY/sL5WT/B2dOciZR2XsGGyPlkeS7wKfpPOI3NeBrwL/CuwG/hh4FbilqmbfeD4jJflL4D+AA/z/3PJX6NxHGNYx/zmdG4mr6HzI211V/5jkQoZ0zN3alNE/VNWnh33MSf6UzlUBdKb8/6WqdvRr3CMbCJKkk43qlJEkaRYDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJav4PHjnTB0Rh5EsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "plt.bar(range(len(h)),h)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index of worst quality : 111\n",
      "Worst Quality sequence is : A@A@A9@?>A45555@@46?54/5/554+588<<8FCBF#############################################################\n"
     ]
    }
   ],
   "source": [
    "def max_poor_qual_cycle(qualities):  #to find which sequencing cycle has the worst quality\n",
    "    min_score = 99999999999\n",
    "    max_index = -1\n",
    "    for i, qual in enumerate(qualities):\n",
    "        score = sum(map(ord,qual))\n",
    "        if min_score > score:\n",
    "            min_score = score\n",
    "            min_index = i\n",
    "    return min_index\n",
    "offset = max_poor_qual_cycle(qualities)\n",
    "print(\"Index of worst quality : %d\" %(offset))\n",
    "print(\"Worst Quality sequence is : \" + qualities[111])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
